### PR TITLE
ARROW-16026: [C++] Add support for the serial executor to expose an async generator as an iterable

### PR DESCRIPTION
**Pull request by westonpace**

### CODE REVIEW

**Comment by westonpace** [comment]
This is based on ARROW-15968 / #12662 and will remain in draft until that PR has merged.
This is needed to enable ARROW-15732 / #12468

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
arrow-s3fs-test failure is unrelated, see ARROW-16043. It seems the Python `test_write_dataset_s3` failure is similar.

File: cpp/src/arrow/util/thread_pool.h
Comment (pitrou) [inline cpp/src/arrow/util/thread_pool.h:281]: Nit, but `SerialExecutor::RunGeneratorInSerialExecutor` seems bit verbose. Perhaps something like `SerialExecutor::IterateGenerator`?

File: cpp/src/arrow/util/thread_pool.h
Comment (pitrou) [inline cpp/src/arrow/util/thread_pool.h:280]: Add a small docstring here?

File: cpp/src/arrow/testing/async_test_util.h
Comment (pitrou) [inline cpp/src/arrow/testing/async_test_util.h:36]: Nit: `FailsAt` sounds like a predicate, perhaps `FailAt`?

File: cpp/src/arrow/util/thread_pool_test.cc
Comment (pitrou) [inline cpp/src/arrow/util/thread_pool_test.cc:310]:
```suggestion
  Executor* captured_executor = nullptr;
```

File: cpp/src/arrow/util/thread_pool_test.cc
Comment (pitrou) [inline cpp/src/arrow/util/thread_pool_test.cc:338]: The "unlike" looks a bit confusing to me, this seems to imply that other follow-up tasks may not be run before the iterator is finished.

File: cpp/src/arrow/util/thread_pool.h
Comment (westonpace) [reply cpp/src/arrow/util/thread_pool.h]: I like that.  I've updated the name.

File: cpp/src/arrow/util/thread_pool.h
Comment (westonpace) [reply cpp/src/arrow/util/thread_pool.h:288]: Added a description

File: cpp/src/arrow/testing/async_test_util.h
Comment (westonpace) [reply cpp/src/arrow/testing/async_test_util.h]: Renamed.

File: cpp/src/arrow/util/thread_pool_test.cc
Comment (westonpace) [reply cpp/src/arrow/util/thread_pool_test.cc]: Done.

File: cpp/src/arrow/util/thread_pool_test.cc
Comment (westonpace) [reply cpp/src/arrow/util/thread_pool_test.cc]: I see your point.  The wording is very ambiguous.  Usually "remaining tasks" get run as part of the following call to `Next`.  Since the last item doesn't have a final call to `Next` we have to run those tasks before we return the final item.  I've updated the comment to just point out what is being tested and avoid making a comparison.

**Pull request closed without merging**
